src/lib/optimizer/optimizer.cpp

```diff
@@ -141,13 +141,23 @@ void Optimizer::add_rule(std::unique_ptr<AbstractRule> rule) {
   _rules.emplace_back(std::move(rule));
 }
 
-std::shared_ptr<AbstractLQPNode> Optimizer::optimize(const std::shared_ptr<AbstractLQPNode>& input) const {
+std::shared_ptr<AbstractLQPNode> Optimizer::optimize(std::shared_ptr<AbstractLQPNode> input) const {
+  // We cannot allow multiple owners of the LQP as one owner could decide to optimize the plan and others might hold a
+  // pointer to a node that is not even part of the plan anymore after optimization. Thus, callers of this method need
+  // to relinquish their ownership (i.e., move their shared_ptr into the method) and take ownership of the resulting
+  // optimized plan.
+  Assert(input.use_count() == 1, "Optimizer should have exclusive ownership of plan");
```

Comment:
Hm, are there any other ways than checking the use_count of the shared pointer to achieve this? This does not seem very clean.

Reply:
Could you try to explain your concerns?

Reply:
As discussed offline

Reply:
Did we reach a conclusion?

Reply:
It seems to be fine.